bson/src/main/org/bson/codecs/pojo/CreatorExecutable.java

```diff
@@ -34,20 +35,20 @@
     private final Class<T> clazz;
     private final Constructor<T> constructor;
     private final Method method;
-    private final List<BsonProperty> properties = new ArrayList<BsonProperty>();
+    private final List<String> properties = new ArrayList<>();
     private final Integer idPropertyIndex;
     private final List<Class<?>> parameterTypes = new ArrayList<Class<?>>();
     private final List<Type> parameterGenericTypes = new ArrayList<Type>();
 
-    CreatorExecutable(final Class<T> clazz, final Constructor<T> constructor) {
-        this(clazz, constructor, null);
+    CreatorExecutable(final Class<T> clazz, final Constructor<T> constructor, final boolean isRecord) {
+        this(clazz, constructor, null, isRecord);
     }
 
-    CreatorExecutable(final Class<T> clazz, final Method method) {
-        this(clazz, null, method);
+    CreatorExecutable(final Class<T> clazz, final Method method, final boolean isRecord) {
+        this(clazz, null, method, isRecord);
     }
 
-    private CreatorExecutable(final Class<T> clazz, final Constructor<T> constructor, final Method method) {
+    private CreatorExecutable(final Class<T> clazz, final Constructor<T> constructor, final Method method, final boolean isRecord) {
```

Comment:
This is starting to looks a little complex - do you think this could be simplified more?

I'm not sure if there is a code smell at the InstanceCreatorFactory level as I can't recall why CreatorExecutable instances were required. Perhaps the abstraction for the implementation of InstanceCreators is not the best? Or alternatively perhaps an interface would be better for CreatorExecutables and then you can separate out and specialize the logic for pojos and records into discrete implementations.

Could you have a look and see if there is an opportunity to simplify here?

Reply:
Yup I'll think about it.

Reply:
@rozza I ended up changing CreatorExecutable to an interface and then creating ConstructorCreatorExecutable, MethodCreatorExecutable, and RecordCreatorExecutable. I think this makes the code pretty clean, but there is a good amount of duplication between Constructor/MethodCreatorExecutable. I can also make a ClassCreatorExecutable which would represent both of them (basically the original CreatorExecutable).